// @mtcute/dispatcher/state/service.js
import { LruMap, timers } from "@fuman/utils";
import { asyncResettable } from "@mtcute/core/utils.js";
const makeCurrentSceneKey = (key) => `$current_scene_${key}`;
class StateService {
  constructor(provider) {
    this.provider = provider;
  }
  _cache = new LruMap(100);
  _vacuumTimer;
  _loaded = false;
  _load = asyncResettable(async () => {
    await this.provider.driver.load?.();
    this._loaded = true;
  });
  async load() {
    await this._load.run();
    this._vacuumTimer = timers.setInterval(() => {
      Promise.resolve(this.provider.state.vacuum(Date.now())).catch(() => {
      });
    }, 3e5);
  }
  async destroy() {
    await this.provider.driver.save?.();
    await this.provider.driver.destroy?.();
    timers.clearInterval(this._vacuumTimer);
    this._loaded = false;
  }
  async getState(key) {
    if (!this._loaded) await this.load();
    const cached = this._cache.get(key);
    if (cached) return cached;
    const state = await this.provider.state.getState(key, Date.now());
    if (!state) return null;
    return JSON.parse(state);
  }
  async setState(key, state, ttl) {
    if (!this._loaded) await this.load();
    this._cache.set(key, state);
    await this.provider.state.setState(key, JSON.stringify(state), ttl);
  }
  async deleteState(key) {
    if (!this._loaded) await this.load();
    this._cache.delete(key);
    await this.provider.state.deleteState(key);
  }
  getCurrentScene(key) {
    return this.getState(makeCurrentSceneKey(key));
  }
  setCurrentScene(key, scene, ttl) {
    return this.setState(makeCurrentSceneKey(key), scene, ttl);
  }
  deleteCurrentScene(key) {
    return this.deleteState(makeCurrentSceneKey(key));
  }
  getRateLimit(key, limit, window) {
    return this.provider.state.getRateLimit(key, Date.now(), limit, window);
  }
  resetRateLimit(key) {
    return this.provider.state.resetRateLimit(key);
  }
}
export {
  StateService
};
