// @mtcute/dispatcher/state/update-state.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const utils = require("@fuman/utils");
const core = require("@mtcute/core");
class RateLimitError extends core.MtcuteError {
  constructor(reset) {
    super("You are being rate limited.");
    this.reset = reset;
  }
}
class UpdateState {
  _key;
  _localKey;
  _storage;
  _scene;
  _scoped;
  _cached;
  _localStorage;
  _localKeyBase;
  constructor(storage, key, scene, scoped, customStorage, customKey) {
    this._storage = storage;
    this._key = key;
    this._scene = scene;
    this._scoped = scoped;
    this._localStorage = customStorage ?? storage;
    this._localKeyBase = customKey ?? key;
    this._updateLocalKey();
  }
  /** Name of the current scene */
  get scene() {
    return this._scene;
  }
  _updateLocalKey() {
    if (!this._scoped) {
      this._localKey = this._localKeyBase;
    } else {
      this._localKey = this._scene ? `${this._scene}_${this._localKeyBase}` : this._localKeyBase;
    }
  }
  async get(fallback, force) {
    if (typeof fallback === "boolean") {
      force = fallback;
      fallback = void 0;
    }
    if (!force && this._cached !== void 0) {
      if (!this._cached && fallback) {
        return typeof fallback === "function" ? fallback() : fallback;
      }
      return this._cached;
    }
    let res = await this._localStorage.getState(this._localKey);
    if (!res && fallback) {
      res = typeof fallback === "function" ? fallback() : fallback;
    }
    this._cached = res;
    return res;
  }
  /**
   * Set new state to the storage
   *
   * @param state  New state
   * @param ttl  TTL for the new state (in seconds)
   */
  async set(state, ttl) {
    this._cached = state;
    await this._localStorage.setState(this._localKey, state, ttl);
  }
  /**
   * Merge the given object to the current state.
   *
   * > **Note**: If the storage currently has no state,
   * > then `fallback` must be provided.
   *
   * Basically a shorthand to calling `.get()`,
   * modifying and then calling `.set()`
   *
   * @param state  State to be merged
   * @param fallback  Default state
   * @param ttl  TTL for the new state (in seconds)
   * @param forceLoad  Whether to force load the old state from storage
   */
  async merge(state, params = {}) {
    const { fallback, ttl, forceLoad } = params;
    const old = await this.get(forceLoad);
    if (!old) {
      if (!fallback) {
        throw new core.MtArgumentError("Cannot use merge on empty state without fallback.");
      }
      const fallback_ = typeof fallback === "function" ? fallback() : fallback;
      await this.set({ ...fallback_, ...state }, ttl);
    } else {
      await this.set({ ...old, ...state }, ttl);
    }
    return this._cached;
  }
  /**
   * Delete the state from the storage
   */
  async delete() {
    this._cached = null;
    await this._localStorage.deleteState(this._localKey);
  }
  /**
   * Enter some scene
   */
  async enter(scene, params) {
    const { with: with_, ttl, reset = true } = params ?? {};
    if (reset && this._scoped) await this.delete();
    if (!scene["_scene"]) {
      throw new core.MtArgumentError("Cannot enter a non-scene Dispatcher");
    }
    if (!scene["_parent"]) {
      throw new core.MtArgumentError("This scene has not been registered");
    }
    this._scene = scene["_scene"];
    this._scoped = scene["_sceneScoped"];
    this._updateLocalKey();
    await this._storage.setCurrentScene(this._key, this._scene, ttl);
    if (with_) {
      if (scene["_customStateKeyDelegate"]) {
        throw new core.MtArgumentError("Cannot use `with` parameter when the scene uses a custom state key delegate");
      }
      await scene.getState(this._key).set(with_, ttl);
    }
  }
  /**
   * Exit from current scene to the root
   *
   * @param reset
   *     Whether to reset scene state (only applicable in case this is a scoped scene)
   */
  async exit(reset = true) {
    if (reset && this._scoped) await this.delete();
    this._scene = null;
    this._updateLocalKey();
    await this._storage.deleteCurrentScene(this._key);
  }
  /**
   * Rate limit some handler.
   *
   * When the rate limit exceeds, {@link RateLimitError} is thrown.
   *
   * This is a simple rate-limiting solution that uses
   * the same key as the state. If you need something more
   * sophisticated and/or customizable, you'll have to implement
   * your own rate-limiter.
   *
   * > **Note**: `key` is used to prefix the local key
   * > derived using the given key delegate.
   *
   * @param key  Key of the rate limit
   * @param limit  Maximum number of requests in `window`
   * @param window  Window size in seconds
   * @returns  Tuple containing the number of remaining and
   *   unix time in ms when the user can try again
   */
  async rateLimit(key, limit, window) {
    const [remaining, reset] = await this._localStorage.getRateLimit(`${key}:${this._localKey}`, limit, window);
    if (!remaining) {
      throw new RateLimitError(reset);
    }
    return [remaining - 1, reset];
  }
  /**
   * Throttle some handler.
   *
   * When the rate limit exceeds, this function waits for it to reset.
   *
   * This is a simple wrapper over {@link rateLimit}, and follows the same logic.
   *
   * > **Note**: `key` is used to prefix the local key
   * > derived using the given key delegate.
   *
   * @param key  Key of the rate limit
   * @param limit  Maximum number of requests in `window`
   * @param window  Window size in seconds
   * @returns  Tuple containing the number of remaining and
   *   unix time in ms when the user can try again
   */
  async throttle(key, limit, window) {
    try {
      return await this.rateLimit(key, limit, window);
    } catch (e) {
      if (e instanceof RateLimitError) {
        await utils.sleep(e.reset - Date.now());
        return this.throttle(key, limit, window);
      }
      throw e;
    }
  }
  /**
   * Reset the rate limit
   *
   * @param key  Key of the rate limit
   */
  async resetRateLimit(key) {
    await this._localStorage.resetRateLimit(`${key}:${this._localKey}`);
  }
}
exports.RateLimitError = RateLimitError;
exports.UpdateState = UpdateState;
