// @mtcute/dispatcher/filters/bots.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const chat = require("./chat.cjs");
const logic = require("./logic.cjs");
function command(commands, {
  prefixes = "/",
  caseSensitive = false
} = {}) {
  if (!Array.isArray(commands)) commands = [commands];
  if (!caseSensitive) {
    commands = commands.map((i) => typeof i === "string" ? i.toLowerCase() : i);
  }
  const argumentsRe = /(["'])(.*?)(?<!\\)\1|(\S+)/g;
  const unescapeRe = /\\(['"])/;
  const commandsRe = [];
  commands.forEach((cmd) => {
    if (typeof cmd !== "string") cmd = cmd.source;
    commandsRe.push(new RegExp(`^(${cmd})(?:\\s|$|@([a-zA-Z0-9_]+?bot)(?:\\s|$))`, caseSensitive ? "" : "i"));
  });
  if (prefixes === null) prefixes = [];
  if (typeof prefixes === "string") prefixes = [prefixes];
  const _prefixes = prefixes;
  const check = (msg) => {
    if (msg.isMessageGroup) return check(msg.messages[0]);
    for (const pref of _prefixes) {
      if (!msg.text.startsWith(pref)) continue;
      const withoutPrefix = msg.text.slice(pref.length);
      for (const regex of commandsRe) {
        const m = withoutPrefix.match(regex);
        if (!m) continue;
        const lastGroup = m[m.length - 1];
        if (lastGroup) {
          const self = msg.client.storage.self.getCached();
          if (self && self.isBot && !self.usernames.some((u) => u.toLowerCase() === lastGroup.toLowerCase())) {
            return false;
          }
        }
        const match = m.slice(1, -1);
        if (!caseSensitive) match[0] = match[0].toLowerCase();
        withoutPrefix.slice(m[0].length).replace(argumentsRe, ($0, $1, $2, $3) => {
          match.push(($2 || $3 || "").replace(unescapeRe, "$1"));
          return "";
        });
        msg.command = match;
        return true;
      }
    }
    return false;
  };
  return check;
}
const start = logic.and(chat.chat("user"), command("start"));
const startGroup = logic.and(logic.or(chat.chat("supergroup"), chat.chat("group")), command("start"));
function deeplinkBase(base) {
  return (params) => {
    if (!Array.isArray(params)) {
      return logic.and(start, (_msg) => {
        const msg = _msg;
        if (msg.command.length !== 2) return false;
        const p = msg.command[1];
        if (typeof params === "string" && p === params) return true;
        const m = p.match(params);
        if (!m) return false;
        msg.command.push(...m.slice(1));
        return true;
      });
    }
    return logic.and(base, (_msg) => {
      const msg = _msg;
      if (msg.command.length !== 2) return false;
      const p = msg.command[1];
      for (const param of params) {
        if (typeof param === "string" && p === param) return true;
        const m = p.match(param);
        if (!m) continue;
        msg.command.push(...m.slice(1));
        return true;
      }
      return false;
    });
  };
}
const deeplink = deeplinkBase(start);
const deeplinkGroup = deeplinkBase(startGroup);
exports.command = command;
exports.deeplink = deeplink;
exports.deeplinkGroup = deeplinkGroup;
exports.start = start;
exports.startGroup = startGroup;
