// @mtcute/dispatcher/filters/bots.js
import { chat } from "./chat.js";
import { and, or } from "./logic.js";
function command(commands, {
  prefixes = "/",
  caseSensitive = false
} = {}) {
  if (!Array.isArray(commands)) commands = [commands];
  if (!caseSensitive) {
    commands = commands.map((i) => typeof i === "string" ? i.toLowerCase() : i);
  }
  const argumentsRe = /(["'])(.*?)(?<!\\)\1|(\S+)/g;
  const unescapeRe = /\\(['"])/;
  const commandsRe = [];
  commands.forEach((cmd) => {
    if (typeof cmd !== "string") cmd = cmd.source;
    commandsRe.push(new RegExp(`^(${cmd})(?:\\s|$|@([a-zA-Z0-9_]+?bot)(?:\\s|$))`, caseSensitive ? "" : "i"));
  });
  if (prefixes === null) prefixes = [];
  if (typeof prefixes === "string") prefixes = [prefixes];
  const _prefixes = prefixes;
  const check = (msg) => {
    if (msg.isMessageGroup) return check(msg.messages[0]);
    for (const pref of _prefixes) {
      if (!msg.text.startsWith(pref)) continue;
      const withoutPrefix = msg.text.slice(pref.length);
      for (const regex of commandsRe) {
        const m = withoutPrefix.match(regex);
        if (!m) continue;
        const lastGroup = m[m.length - 1];
        if (lastGroup) {
          const self = msg.client.storage.self.getCached();
          if (self && self.isBot && !self.usernames.some((u) => u.toLowerCase() === lastGroup.toLowerCase())) {
            return false;
          }
        }
        const match = m.slice(1, -1);
        if (!caseSensitive) match[0] = match[0].toLowerCase();
        withoutPrefix.slice(m[0].length).replace(argumentsRe, ($0, $1, $2, $3) => {
          match.push(($2 || $3 || "").replace(unescapeRe, "$1"));
          return "";
        });
        msg.command = match;
        return true;
      }
    }
    return false;
  };
  return check;
}
const start = and(chat("user"), command("start"));
const startGroup = and(or(chat("supergroup"), chat("group")), command("start"));
function deeplinkBase(base) {
  return (params) => {
    if (!Array.isArray(params)) {
      return and(start, (_msg) => {
        const msg = _msg;
        if (msg.command.length !== 2) return false;
        const p = msg.command[1];
        if (typeof params === "string" && p === params) return true;
        const m = p.match(params);
        if (!m) return false;
        msg.command.push(...m.slice(1));
        return true;
      });
    }
    return and(base, (_msg) => {
      const msg = _msg;
      if (msg.command.length !== 2) return false;
      const p = msg.command[1];
      for (const param of params) {
        if (typeof param === "string" && p === param) return true;
        const m = p.match(param);
        if (!m) continue;
        msg.command.push(...m.slice(1));
        return true;
      }
      return false;
    });
  };
}
const deeplink = deeplinkBase(start);
const deeplinkGroup = deeplinkBase(startGroup);
export {
  command,
  deeplink,
  deeplinkGroup,
  start,
  startGroup
};
