// @mtcute/dispatcher/filters/bundle.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const bots = require("./bots.cjs");
const chat = require("./chat.cjs");
const group = require("./group.cjs");
const logic = require("./logic.cjs");
const message = require("./message.cjs");
const state = require("./state.cjs");
const text = require("./text.cjs");
const updates = require("./updates.cjs");
const user = require("./user.cjs");
exports.command = bots.command;
exports.deeplink = bots.deeplink;
exports.deeplinkGroup = bots.deeplinkGroup;
exports.start = bots.start;
exports.startGroup = bots.startGroup;
exports.chat = chat.chat;
exports.chatId = chat.chatId;
exports.every = group.every;
exports.some = group.some;
exports.and = logic.and;
exports.any = logic.any;
exports.not = logic.not;
exports.or = logic.or;
exports.action = message.action;
exports.animation = message.animation;
exports.anyDocument = message.anyDocument;
exports.anyLocation = message.anyLocation;
exports.anyVideo = message.anyVideo;
exports.audio = message.audio;
exports.contact = message.contact;
exports.dice = message.dice;
exports.document = message.document;
exports.game = message.game;
exports.incoming = message.incoming;
exports.invoice = message.invoice;
exports.liveLocation = message.liveLocation;
exports.location = message.location;
exports.media = message.media;
exports.mediaOf = message.mediaOf;
exports.outgoing = message.outgoing;
exports.photo = message.photo;
exports.poll = message.poll;
exports.reply = message.reply;
exports.replyOrigin = message.replyOrigin;
exports.replyTo = message.replyTo;
exports.roundMessage = message.roundMessage;
exports.scheduled = message.scheduled;
exports.sender = message.sender;
exports.service = message.service;
exports.sticker = message.sticker;
exports.stickerBySourceType = message.stickerBySourceType;
exports.stickerByType = message.stickerByType;
exports.story = message.story;
exports.text = message.text;
exports.venue = message.venue;
exports.video = message.video;
exports.voice = message.voice;
exports.webpage = message.webpage;
exports.withCompleteChat = message.withCompleteChat;
exports.withCompleteSender = message.withCompleteSender;
exports.state = state.state;
exports.stateEmpty = state.stateEmpty;
exports.contains = text.contains;
exports.endsWith = text.endsWith;
exports.equals = text.equals;
exports.regex = text.regex;
exports.startsWith = text.startsWith;
exports.chatMember = updates.chatMember;
exports.chatMemberSelf = updates.chatMemberSelf;
exports.userStatus = updates.userStatus;
exports.bot = user.bot;
exports.me = user.me;
exports.userId = user.userId;
