// @mtcute/dispatcher/filters/message.js
import { RawLocation, RawDocument } from "@mtcute/core";
const incoming = (msg) => !msg.isOutgoing;
const outgoing = (msg) => msg.isOutgoing;
const scheduled = (msg) => msg.isScheduled;
const reply = (msg) => msg.replyToMessage !== null;
function replyOrigin(origin) {
  return (msg) => msg.replyToMessage?.originIs(origin) ?? false;
}
const media = (msg) => msg.media !== null;
function mediaOf(type) {
  return (msg) => msg.media?.type === type;
}
const photo = mediaOf("photo");
const dice = mediaOf("dice");
const contact = mediaOf("contact");
const audio = mediaOf("audio");
const voice = mediaOf("voice");
const sticker = mediaOf("sticker");
const document = mediaOf("document");
const anyVideo = mediaOf("video");
const location = mediaOf("location");
const liveLocation = mediaOf("live_location");
const game = mediaOf("game");
const webpage = mediaOf("webpage");
const venue = mediaOf("venue");
const poll = mediaOf("poll");
const invoice = mediaOf("invoice");
const story = mediaOf("story");
const anyLocation = (msg) => msg.media instanceof RawLocation;
const anyDocument = (msg) => msg.media instanceof RawDocument;
const video = (msg) => msg.media?.type === "video" && !msg.media.isAnimation && !msg.media.isRound;
const animation = (msg) => msg.media?.type === "video" && msg.media.isAnimation && !msg.media.isRound;
const roundMessage = (msg) => msg.media?.type === "video" && !msg.media.isAnimation && msg.media.isRound;
function stickerByType(type) {
  return (msg) => msg.media?.type === "sticker" && msg.media.stickerType === type;
}
function stickerBySourceType(type) {
  return (msg) => msg.media?.type === "sticker" && msg.media.sourceType === type;
}
const text = (msg) => msg.media === null && !msg.isService;
const service = (msg) => msg.isService;
function action(type) {
  if (Array.isArray(type)) {
    const index = {};
    type.forEach((it) => index[it] = true);
    return (msg) => msg.action?.type in index;
  }
  return (msg) => msg.action?.type === type;
}
function sender(type) {
  return (msg) => msg.sender.type === type;
}
function replyTo(filter) {
  return async (msg, state) => {
    if (!msg.replyToMessage?.id) return false;
    const reply2 = msg._name === "new_message" ? await msg.getReplyTo() : msg.replyTo;
    if (!reply2) return false;
    if (msg._name === "new_message") {
      msg.getReplyTo = () => Promise.resolve(reply2);
    }
    if (!filter) return true;
    return filter(reply2, state);
  };
}
function withCompleteSender(filter) {
  return async (msg, state) => {
    try {
      await msg.getCompleteSender();
    } catch {
      return false;
    }
    if (!filter) return true;
    return filter(msg, state);
  };
}
function withCompleteChat(filter) {
  return async (msg, state) => {
    try {
      await msg.getCompleteChat();
    } catch {
      return false;
    }
    if (!filter) return true;
    return filter(msg, state);
  };
}
export {
  action,
  animation,
  anyDocument,
  anyLocation,
  anyVideo,
  audio,
  contact,
  dice,
  document,
  game,
  incoming,
  invoice,
  liveLocation,
  location,
  media,
  mediaOf,
  outgoing,
  photo,
  poll,
  reply,
  replyOrigin,
  replyTo,
  roundMessage,
  scheduled,
  sender,
  service,
  sticker,
  stickerBySourceType,
  stickerByType,
  story,
  text,
  venue,
  video,
  voice,
  webpage,
  withCompleteChat,
  withCompleteSender
};
