// @mtcute/dispatcher/filters/text.js
import { Message, BusinessMessage } from "@mtcute/core";
function extractText(obj) {
  if (obj instanceof Message || obj instanceof BusinessMessage) {
    return obj.text;
  }
  switch (obj._name) {
    case "inline_query":
      return obj.query;
    case "chosen_inline_result":
      return obj.id;
    case "callback_query":
    case "inline_callback_query":
    case "business_callback_query":
      if (obj.raw.data) return obj.dataStr;
  }
  return null;
}
function regex(regex2) {
  return (obj) => {
    const txt = extractText(obj);
    if (!txt) return false;
    const m = txt.match(regex2);
    if (m) {
      obj.match = m;
      return true;
    }
    return false;
  };
}
function equals(str, ignoreCase = false) {
  if (ignoreCase) {
    str = str.toLowerCase();
    return (obj) => extractText(obj)?.toLowerCase() === str;
  }
  return (obj) => extractText(obj) === str;
}
function contains(str, ignoreCase = false) {
  if (ignoreCase) {
    str = str.toLowerCase();
    return (obj) => {
      const txt = extractText(obj);
      return txt != null && txt.toLowerCase().includes(str);
    };
  }
  return (obj) => {
    const txt = extractText(obj);
    return txt != null && txt.includes(str);
  };
}
function startsWith(str, ignoreCase = false) {
  if (ignoreCase) {
    str = str.toLowerCase();
    return (obj) => {
      const txt = extractText(obj);
      return txt != null && txt.toLowerCase().substring(0, str.length) === str;
    };
  }
  return (obj) => {
    const txt = extractText(obj);
    return txt != null && txt.substring(0, str.length) === str;
  };
}
function endsWith(str, ignoreCase = false) {
  if (ignoreCase) {
    str = str.toLowerCase();
    return (obj) => {
      const txt = extractText(obj);
      return txt != null && txt.toLowerCase().substring(0, str.length) === str;
    };
  }
  return (obj) => {
    const txt = extractText(obj);
    return txt != null && txt.substring(0, str.length) === str;
  };
}
export {
  contains,
  endsWith,
  equals,
  regex,
  startsWith
};
