// @mtcute/dispatcher/filters/text.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const core = require("@mtcute/core");
function extractText(obj) {
  if (obj instanceof core.Message || obj instanceof core.BusinessMessage) {
    return obj.text;
  }
  switch (obj._name) {
    case "inline_query":
      return obj.query;
    case "chosen_inline_result":
      return obj.id;
    case "callback_query":
    case "inline_callback_query":
    case "business_callback_query":
      if (obj.raw.data) return obj.dataStr;
  }
  return null;
}
function regex(regex2) {
  return (obj) => {
    const txt = extractText(obj);
    if (!txt) return false;
    const m = txt.match(regex2);
    if (m) {
      obj.match = m;
      return true;
    }
    return false;
  };
}
function equals(str, ignoreCase = false) {
  if (ignoreCase) {
    str = str.toLowerCase();
    return (obj) => extractText(obj)?.toLowerCase() === str;
  }
  return (obj) => extractText(obj) === str;
}
function contains(str, ignoreCase = false) {
  if (ignoreCase) {
    str = str.toLowerCase();
    return (obj) => {
      const txt = extractText(obj);
      return txt != null && txt.toLowerCase().includes(str);
    };
  }
  return (obj) => {
    const txt = extractText(obj);
    return txt != null && txt.includes(str);
  };
}
function startsWith(str, ignoreCase = false) {
  if (ignoreCase) {
    str = str.toLowerCase();
    return (obj) => {
      const txt = extractText(obj);
      return txt != null && txt.toLowerCase().substring(0, str.length) === str;
    };
  }
  return (obj) => {
    const txt = extractText(obj);
    return txt != null && txt.substring(0, str.length) === str;
  };
}
function endsWith(str, ignoreCase = false) {
  if (ignoreCase) {
    str = str.toLowerCase();
    return (obj) => {
      const txt = extractText(obj);
      return txt != null && txt.toLowerCase().substring(0, str.length) === str;
    };
  }
  return (obj) => {
    const txt = extractText(obj);
    return txt != null && txt.substring(0, str.length) === str;
  };
}
exports.contains = contains;
exports.endsWith = endsWith;
exports.equals = equals;
exports.regex = regex;
exports.startsWith = startsWith;
