// @mtcute/dispatcher/callback-data-builder.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const core = require("@mtcute/core");
class CallbackDataBuilder {
  /**
   * @param prefix  Prefix for the data. Use something unique across your bot.
   * @param fields  Field names in the order they will be serialized.
   */
  constructor(prefix, ...fields) {
    this.prefix = prefix;
    this._fields = fields;
  }
  _fields;
  sep = ":";
  /**
   * Build a callback data string
   *
   * @param obj  Object containing the data
   */
  build(obj) {
    const ret = this.prefix + this.sep + this._fields.map((f) => {
      const val = obj[f];
      if (val.includes(this.sep)) {
        throw new core.MtArgumentError(
          `Value for ${f} ${val} contains separator ${this.sep} and cannot be used.`
        );
      }
      return val;
    }).join(this.sep);
    if (ret.length > 64) {
      throw new core.MtArgumentError("Resulting callback data is too long.");
    }
    return ret;
  }
  parse(data, safe = false) {
    const parts = data.split(this.sep);
    if (parts[0] !== this.prefix) {
      if (safe) return null;
      throw new core.MtArgumentError(
        `Invalid data passed: "${data}" (bad prefix, expected ${this.prefix}, got ${parts[0]})`
      );
    }
    if (parts.length !== this._fields.length + 1) {
      if (safe) return null;
      throw new core.MtArgumentError(
        `Invalid data passed: "${data}" (bad parts count, expected ${this._fields.length}, got ${parts.length - 1})`
      );
    }
    const ret = {};
    parts.forEach((it, idx) => {
      if (idx === 0) return;
      ret[this._fields[idx - 1]] = it;
    });
    return ret;
  }
  /**
   * Create a filter for this callback data.
   *
   * You can either pass an object with field names as keys and values as strings or regexes,
   * which will be compiled to a RegExp, or a function that will be called with the parsed data.
   * Note that the strings will be passed to `RegExp` **directly**, so you may want to escape them.
   *
   * When using a function, you can either return a boolean, or an object with field names as keys
   * and values as strings or regexes. In the latter case, the resulting object will be matched
   * against the parsed data the same way as if you passed it directly.
   *
   * @param params
   */
  filter(params = {}) {
    if (typeof params === "function") {
      return async (query) => {
        if (!query.dataStr) return false;
        const data = this.parse(query.dataStr, true);
        if (!data) return false;
        const fnResult = await params(query, data);
        if (typeof fnResult === "boolean") {
          query.match = data;
          return fnResult;
        }
        for (const key in fnResult) {
          const value = data[key];
          if (value === void 0) return false;
          let matchers = fnResult[key];
          if (!Array.isArray(matchers)) matchers = [matchers];
          for (const matcher of matchers) {
            if (typeof matcher === "string") {
              if (value !== matcher) return false;
            } else if (!matcher.test(value)) {
              return false;
            }
          }
        }
        query.match = data;
        return true;
      };
    }
    const parts = [];
    this._fields.forEach((field) => {
      if (!(field in params)) {
        parts.push(`[^${this.sep}]*?`);
        return;
      }
      const value = params[field];
      if (Array.isArray(value)) {
        parts.push(`(${value.map((i) => typeof i === "string" ? i : i.source).join("|")})`);
      } else {
        parts.push(typeof value === "string" ? value : value.source);
      }
    });
    const regex = new RegExp(`^${this.prefix}${this.sep}${parts.join(this.sep)}$`);
    return (query) => {
      const m = query.dataStr?.match(regex);
      if (!m) {
        return false;
      }
      query.match = this.parse(m[0]);
      return true;
    };
  }
}
exports.CallbackDataBuilder = CallbackDataBuilder;
