// @mtcute/dispatcher/index.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const callbackDataBuilder = require("./callback-data-builder.cjs");
const dispatcher = require("./dispatcher.cjs");
const propagation = require("./propagation.cjs");
const wizard = require("./wizard.cjs");
const botGuestChatQuery = require("./context/bot-guest-chat-query.cjs");
const businessMessage = require("./context/business-message.cjs");
const callbackQuery = require("./context/callback-query.cjs");
const chatJoinRequest = require("./context/chat-join-request.cjs");
const chosenInlineResult = require("./context/chosen-inline-result.cjs");
const inlineQuery = require("./context/inline-query.cjs");
const message = require("./context/message.cjs");
const preCheckoutQuery = require("./context/pre-checkout-query.cjs");
const bundle = require("./filters/bundle.cjs");
const key = require("./state/key.cjs");
const memory = require("./state/providers/memory.cjs");
const sqlite = require("./state/providers/sqlite.cjs");
const updateState = require("./state/update-state.cjs");
exports.CallbackDataBuilder = callbackDataBuilder.CallbackDataBuilder;
exports.Dispatcher = dispatcher.Dispatcher;
exports.PropagationAction = propagation.PropagationAction;
exports.WizardScene = wizard.WizardScene;
exports.WizardSceneAction = wizard.WizardSceneAction;
exports.BotGuestChatQueryContext = botGuestChatQuery.BotGuestChatQueryContext;
exports.BusinessMessageContext = businessMessage.BusinessMessageContext;
exports.BusinessCallbackQueryContext = callbackQuery.BusinessCallbackQueryContext;
exports.CallbackQueryContext = callbackQuery.CallbackQueryContext;
exports.InlineCallbackQueryContext = callbackQuery.InlineCallbackQueryContext;
exports.ChatJoinRequestUpdateContext = chatJoinRequest.ChatJoinRequestUpdateContext;
exports.ChosenInlineResultContext = chosenInlineResult.ChosenInlineResultContext;
exports.InlineQueryContext = inlineQuery.InlineQueryContext;
exports.MessageContext = message.MessageContext;
exports.PreCheckoutQueryContext = preCheckoutQuery.PreCheckoutQueryContext;
exports.filters = bundle;
exports.defaultStateKeyDelegate = key.defaultStateKeyDelegate;
exports.MemoryStateStorage = memory.MemoryStateStorage;
exports.SqliteStateStorage = sqlite.SqliteStateStorage;
exports.RateLimitError = updateState.RateLimitError;
exports.UpdateState = updateState.UpdateState;
