// @mtcute/dispatcher/wizard.js
import { Dispatcher } from "./dispatcher.js";
import { or } from "./filters/logic.js";
import { state, stateEmpty } from "./filters/state.js";
var WizardSceneAction = /* @__PURE__ */ ((WizardSceneAction2) => {
  WizardSceneAction2["Next"] = "next";
  WizardSceneAction2["Stay"] = "stay";
  WizardSceneAction2["Exit"] = "exit";
  return WizardSceneAction2;
})(WizardSceneAction || {});
class WizardScene extends Dispatcher {
  _steps = 0;
  _defaultState = {};
  constructor(name, params) {
    super(void 0, { sceneName: name, ...params });
  }
  setDefaultState(defaultState) {
    this._defaultState = defaultState;
  }
  /**
   * Get the total number of registered steps
   */
  get totalSteps() {
    return this._steps;
  }
  /**
   * Go to the Nth step
   */
  async goToStep(state2, step) {
    if (step >= this._steps) {
      await state2.exit();
    } else {
      await state2.merge({ $step: step }, { fallback: this._defaultState });
    }
  }
  /**
   * Skip N steps
   */
  async skip(state2, count = 1) {
    const { $step } = await state2.get() || {};
    if ($step === void 0) throw new Error("Wizard state is not initialized");
    return this.goToStep(state2, $step + count);
  }
  /**
   * Filter that will only pass if the current step is `step`
   */
  // eslint-disable-next-line ts/no-empty-object-type
  static onNthStep(step) {
    if (step === 0) {
      return or(
        stateEmpty,
        state((it) => it.$step === void 0 || it.$step === 0)
      );
    }
    return state((it) => it.$step === step);
  }
  /**
   * Filter that will only pass if the current step is the one after last one added
   */
  // eslint-disable-next-line ts/no-empty-object-type
  onCurrentStep() {
    return WizardScene.onNthStep(this._steps);
  }
  /**
   * Add a step to the wizard
   */
  addStep(handler) {
    const step = this._steps++;
    this.onNewMessage(WizardScene.onNthStep(step), async (msg, state2) => {
      const result = await handler(msg, state2);
      if (typeof result === "number") {
        await this.goToStep(state2, result);
        return;
      }
      switch (result) {
        case "next": {
          await this.goToStep(state2, step + 1);
          break;
        }
        case "exit":
          await state2.exit();
          break;
      }
    });
  }
}
export {
  WizardScene,
  WizardSceneAction
};
