// @mtcute/dispatcher/wizard.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const dispatcher = require("./dispatcher.cjs");
const logic = require("./filters/logic.cjs");
const state = require("./filters/state.cjs");
var WizardSceneAction = /* @__PURE__ */ ((WizardSceneAction2) => {
  WizardSceneAction2["Next"] = "next";
  WizardSceneAction2["Stay"] = "stay";
  WizardSceneAction2["Exit"] = "exit";
  return WizardSceneAction2;
})(WizardSceneAction || {});
class WizardScene extends dispatcher.Dispatcher {
  _steps = 0;
  _defaultState = {};
  constructor(name, params) {
    super(void 0, { sceneName: name, ...params });
  }
  setDefaultState(defaultState) {
    this._defaultState = defaultState;
  }
  /**
   * Get the total number of registered steps
   */
  get totalSteps() {
    return this._steps;
  }
  /**
   * Go to the Nth step
   */
  async goToStep(state2, step) {
    if (step >= this._steps) {
      await state2.exit();
    } else {
      await state2.merge({ $step: step }, { fallback: this._defaultState });
    }
  }
  /**
   * Skip N steps
   */
  async skip(state2, count = 1) {
    const { $step } = await state2.get() || {};
    if ($step === void 0) throw new Error("Wizard state is not initialized");
    return this.goToStep(state2, $step + count);
  }
  /**
   * Filter that will only pass if the current step is `step`
   */
  // eslint-disable-next-line ts/no-empty-object-type
  static onNthStep(step) {
    if (step === 0) {
      return logic.or(
        state.stateEmpty,
        state.state((it) => it.$step === void 0 || it.$step === 0)
      );
    }
    return state.state((it) => it.$step === step);
  }
  /**
   * Filter that will only pass if the current step is the one after last one added
   */
  // eslint-disable-next-line ts/no-empty-object-type
  onCurrentStep() {
    return WizardScene.onNthStep(this._steps);
  }
  /**
   * Add a step to the wizard
   */
  addStep(handler) {
    const step = this._steps++;
    this.onNewMessage(WizardScene.onNthStep(step), async (msg, state2) => {
      const result = await handler(msg, state2);
      if (typeof result === "number") {
        await this.goToStep(state2, result);
        return;
      }
      switch (result) {
        case "next": {
          await this.goToStep(state2, step + 1);
          break;
        }
        case "exit":
          await state2.exit();
          break;
      }
    });
  }
}
exports.WizardScene = WizardScene;
exports.WizardSceneAction = WizardSceneAction;
