// @mtcute/dispatcher/state/providers/memory.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const core = require("@mtcute/core");
class MemoryStateRepository {
  constructor(_driver) {
    this._driver = _driver;
    this.state = this._driver.getState("dispatcher_fsm", () => /* @__PURE__ */ new Map());
    this.rl = this._driver.getState("rl", () => /* @__PURE__ */ new Map());
  }
  state;
  rl;
  setState(key, state, ttl) {
    this.state.set(key, {
      value: state,
      expiresAt: ttl ? Date.now() + ttl * 1e3 : void 0
    });
  }
  getState(key, now) {
    const state = this.state.get(key);
    if (!state) return null;
    if (state.expiresAt && state.expiresAt < now) {
      this.state.delete(key);
      return null;
    }
    return state.value;
  }
  deleteState(key) {
    this.state.delete(key);
  }
  vacuum(now) {
    for (const [key, state] of this.state.entries()) {
      if (state.expiresAt && state.expiresAt < now) {
        this.state.delete(key);
      }
    }
    for (const [key, state] of this.rl.entries()) {
      if (state.reset < now) {
        this.rl.delete(key);
      }
    }
  }
  getRateLimit(key, now, limit, window) {
    const item = this.rl.get(key);
    if (!item) {
      const state = {
        reset: now + window * 1e3,
        remaining: limit
      };
      this.rl.set(key, state);
      return [state.remaining, state.reset];
    }
    if (item.reset < now) {
      const state = {
        reset: now + window * 1e3,
        remaining: limit
      };
      this.rl.set(key, state);
      return [state.remaining, state.reset];
    }
    item.remaining = item.remaining > 0 ? item.remaining - 1 : 0;
    return [item.remaining, item.reset];
  }
  resetRateLimit(key) {
    this.rl.delete(key);
  }
}
class MemoryStateStorage {
  constructor(driver = new core.MemoryStorageDriver()) {
    this.driver = driver;
    this.state = new MemoryStateRepository(this.driver);
  }
  state;
}
exports.MemoryStateStorage = MemoryStateStorage;
