// @mtcute/dispatcher/state/providers/sqlite.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
class SqliteStateRepository {
  constructor(_driver) {
    this._driver = _driver;
    _driver.registerMigration("state", 1, (db) => {
      db.exec(`
          create table fsm_state (
              key text primary key,
              value text not null,
              expires_at integer
          );
          create table rl_state (
              key text primary key,
              reset integer not null,
              remaining integer not null
          );
      `);
    });
    _driver.onLoad(() => {
      this._setState = _driver.db.prepare(
        "insert or replace into fsm_state (key, value, expires_at) values (?, ?, ?)"
      );
      this._getState = _driver.db.prepare("select value, expires_at from fsm_state where key = ?");
      this._deleteState = _driver.db.prepare("delete from fsm_state where key = ?");
      this._deleteOldState = _driver.db.prepare("delete from fsm_state where expires_at < ?");
      this._setRl = _driver.db.prepare("insert or replace into rl_state (key, reset, remaining) values (?, ?, ?)");
      this._getRl = _driver.db.prepare("select reset, remaining from rl_state where key = ?");
      this._deleteRl = _driver.db.prepare("delete from rl_state where key = ?");
      this._deleteOldRl = _driver.db.prepare("delete from rl_state where reset < ?");
    });
  }
  _setState;
  setState(key, state, ttl) {
    this._setState.run(key, state, ttl ? Date.now() + ttl * 1e3 : void 0);
  }
  _getState;
  getState(key, now) {
    const res_ = this._getState.get(key);
    if (!res_) return null;
    const res = res_;
    if (res.expires_at && res.expires_at < now) {
      this._deleteState.run(key);
      return null;
    }
    return res.value;
  }
  _deleteState;
  deleteState(key) {
    this._deleteState.run(key);
  }
  _deleteOldState;
  _deleteOldRl;
  vacuum(now) {
    this._deleteOldState.run(now);
    this._deleteOldRl.run(now);
  }
  _setRl;
  _getRl;
  _deleteRl;
  getRateLimit(key, now, limit, window) {
    const val = this._getRl.get(key);
    if (!val || val.reset < now) {
      const item = {
        reset: now + window * 1e3,
        remaining: limit
      };
      this._setRl.run(key, item.reset, item.remaining);
      return [item.remaining, item.reset];
    }
    if (val.remaining > 0) {
      val.remaining -= 1;
      this._setRl.run(key, val.reset, val.remaining);
    }
    return [val.remaining, val.reset];
  }
  resetRateLimit(key) {
    this._deleteRl.run(key);
  }
}
class SqliteStateStorage {
  constructor(driver) {
    this.driver = driver;
    this.state = new SqliteStateRepository(driver);
  }
  state;
  static from(provider) {
    return new SqliteStateStorage(provider.driver);
  }
}
exports.SqliteStateStorage = SqliteStateStorage;
