// @mtcute/dispatcher/context/business-message.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const core = require("@mtcute/core");
class BusinessMessageContext extends core.BusinessMessage {
  constructor(client, message) {
    const msg = Array.isArray(message) ? message[message.length - 1] : message;
    super(msg.update, msg._peers);
    this.client = client;
    this.messages = Array.isArray(message) ? message.map((it) => new BusinessMessageContext(client, it)) : [this];
    this.isMessageGroup = Array.isArray(message);
  }
  // this is primarily for proper types in filters, so don't bother much with actual value
  _name = "new_business_message";
  /**
   * List of messages in the message group.
   *
   * For other updates, this is a list with a single element (`this`).
   */
  messages;
  /** Whether this update is about a message group */
  isMessageGroup;
  /** Get all custom emojis contained in this message (message group), if any */
  getCustomEmojis() {
    return this.client.getCustomEmojisFromMessages(this.messages);
  }
  /** Send a text message to the same chat (and topic, if applicable) as a given message */
  answerText(...params) {
    const [send, params_ = {}] = params;
    params_.businessConnectionId = this.update.connectionId;
    return this.client.answerText(this, send, params_);
  }
  /** Send a media to the same chat (and topic, if applicable) as a given message */
  answerMedia(...params) {
    const [send, params_ = {}] = params;
    params_.businessConnectionId = this.update.connectionId;
    return this.client.answerMedia(this, send, params_);
  }
  /** Send a media group to the same chat (and topic, if applicable) as a given message */
  answerMediaGroup(...params) {
    const [send, params_ = {}] = params;
    params_.businessConnectionId = this.update.connectionId;
    return this.client.answerMediaGroup(this, send, params_);
  }
  /** Send a text message in reply to this message */
  replyText(...params) {
    const [send, params_ = {}] = params;
    params_.businessConnectionId = this.update.connectionId;
    return this.client.replyText(this, send, params_);
  }
  /** Send a media in reply to this message */
  replyMedia(...params) {
    const [send, params_ = {}] = params;
    params_.businessConnectionId = this.update.connectionId;
    return this.client.replyMedia(this, send, params_);
  }
  /** Send a media group in reply to this message */
  replyMediaGroup(...params) {
    const [send, params_ = {}] = params;
    params_.businessConnectionId = this.update.connectionId;
    return this.client.replyMediaGroup(this, send, params_);
  }
  /** Send a text message in reply to this message */
  quoteWithText(params) {
    params.businessConnectionId = this.update.connectionId;
    return this.client.quoteWithText(this, params);
  }
  /** Send a media in reply to this message */
  quoteWithMedia(params) {
    params.businessConnectionId = this.update.connectionId;
    return this.client.quoteWithMedia(this, params);
  }
  /** Send a media group in reply to this message */
  quoteWithMediaGroup(params) {
    params.businessConnectionId = this.update.connectionId;
    return this.client.quoteWithMediaGroup(this, params);
  }
  /** Delete this message (message group) */
  delete(params) {
    return this.client.deleteMessagesById(
      this.chat.inputPeer,
      this.messages.map((it) => it.id),
      params
    );
  }
  /** Pin this message */
  pin(params) {
    return this.client.pinMessage({
      chatId: this.chat.inputPeer,
      message: this.id,
      ...params
    });
  }
  /** Unpin this message */
  unpin() {
    return this.client.unpinMessage({
      chatId: this.chat.inputPeer,
      message: this.id
    });
  }
  /** Edit this message */
  edit(params) {
    return this.client.editMessage({
      chatId: this.chat.inputPeer,
      message: this.id,
      ...params
    });
  }
  /** Forward this message (message group) */
  forwardTo(params) {
    return this.client.forwardMessagesById({
      fromChatId: this.chat.inputPeer,
      messages: this.messages.map((it) => it.id),
      ...params
    });
  }
  /** Send a copy of this message (message group) */
  copy(params) {
    if (this.isMessageGroup) {
      return this.client.sendCopyGroup({
        messages: this.messages,
        ...params
      });
    }
    return this.client.sendCopy({
      message: this,
      ...params
    });
  }
  /** React to this message */
  react(params) {
    return this.client.sendReaction({
      chatId: this.chat.inputPeer,
      message: this.id,
      ...params
    });
  }
}
exports.BusinessMessageContext = BusinessMessageContext;
