// @mtcute/dispatcher/context/callback-query.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const core = require("@mtcute/core");
class CallbackQueryContext extends core.CallbackQuery {
  constructor(client, query) {
    super(query.raw, query._peers);
    this.client = client;
  }
  _name = "callback_query";
  /** Answer to this callback query */
  answer(params) {
    return this.client.answerCallbackQuery(this.id, params);
  }
  /**
   * Get the message containing the callback button being clicked.
   *
   * Note that the message may have been deleted, in which case
   * `null` will be returned.
   */
  async getMessage() {
    return this.client.getCallbackQueryMessage(this);
  }
  /**
   * Edit the message that contained the callback button that was clicked.
   */
  async editMessage(params) {
    return this.client.editMessage({
      chatId: this.raw.peer,
      message: this.raw.msgId,
      ...params
    });
  }
  /**
   * Shortcut for getting the message and editing it.
   */
  async editMessageWith(handler) {
    const msg = await this.getMessage();
    if (!msg) return;
    const res = await handler(msg);
    if (!res) return;
    return this.editMessage(res);
  }
}
class InlineCallbackQueryContext extends core.InlineCallbackQuery {
  constructor(client, query) {
    super(query.raw, query._peers);
    this.client = client;
  }
  _name = "inline_callback_query";
  /** Answer to this callback query */
  answer(params) {
    return this.client.answerCallbackQuery(this.id, params);
  }
  /**
   * Edit the message that contained the callback button that was clicked.
   */
  async editMessage(params) {
    return this.client.editInlineMessage({
      messageId: this.raw.msgId,
      ...params
    });
  }
}
class BusinessCallbackQueryContext extends core.BusinessCallbackQuery {
  constructor(client, query) {
    super(query.raw, query._peers);
    this.client = client;
  }
  _name = "business_callback_query";
  /** Answer to this callback query */
  answer(params) {
    return this.client.answerCallbackQuery(this.id, params);
  }
  /**
   * Edit the message that contained the callback button that was clicked.
   */
  async editMessage(params) {
    return this.client.editMessage({
      message: this.message,
      businessConnectionId: this.connectionId,
      ...params
    });
  }
}
exports.BusinessCallbackQueryContext = BusinessCallbackQueryContext;
exports.CallbackQueryContext = CallbackQueryContext;
exports.InlineCallbackQueryContext = InlineCallbackQueryContext;
