// @mtcute/dispatcher/context/callback-query.js
import { CallbackQuery, InlineCallbackQuery, BusinessCallbackQuery } from "@mtcute/core";
class CallbackQueryContext extends CallbackQuery {
  constructor(client, query) {
    super(query.raw, query._peers);
    this.client = client;
  }
  _name = "callback_query";
  /** Answer to this callback query */
  answer(params) {
    return this.client.answerCallbackQuery(this.id, params);
  }
  /**
   * Get the message containing the callback button being clicked.
   *
   * Note that the message may have been deleted, in which case
   * `null` will be returned.
   */
  async getMessage() {
    return this.client.getCallbackQueryMessage(this);
  }
  /**
   * Edit the message that contained the callback button that was clicked.
   */
  async editMessage(params) {
    return this.client.editMessage({
      chatId: this.raw.peer,
      message: this.raw.msgId,
      ...params
    });
  }
  /**
   * Shortcut for getting the message and editing it.
   */
  async editMessageWith(handler) {
    const msg = await this.getMessage();
    if (!msg) return;
    const res = await handler(msg);
    if (!res) return;
    return this.editMessage(res);
  }
}
class InlineCallbackQueryContext extends InlineCallbackQuery {
  constructor(client, query) {
    super(query.raw, query._peers);
    this.client = client;
  }
  _name = "inline_callback_query";
  /** Answer to this callback query */
  answer(params) {
    return this.client.answerCallbackQuery(this.id, params);
  }
  /**
   * Edit the message that contained the callback button that was clicked.
   */
  async editMessage(params) {
    return this.client.editInlineMessage({
      messageId: this.raw.msgId,
      ...params
    });
  }
}
class BusinessCallbackQueryContext extends BusinessCallbackQuery {
  constructor(client, query) {
    super(query.raw, query._peers);
    this.client = client;
  }
  _name = "business_callback_query";
  /** Answer to this callback query */
  answer(params) {
    return this.client.answerCallbackQuery(this.id, params);
  }
  /**
   * Edit the message that contained the callback button that was clicked.
   */
  async editMessage(params) {
    return this.client.editMessage({
      message: this.message,
      businessConnectionId: this.connectionId,
      ...params
    });
  }
}
export {
  BusinessCallbackQueryContext,
  CallbackQueryContext,
  InlineCallbackQueryContext
};
