// @mtcute/dispatcher/context/chat-join-request.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const core = require("@mtcute/core");
class ChatJoinRequestUpdateContext extends core.BotChatJoinRequestUpdate {
  constructor(client, update) {
    super(update.raw, update._peers);
    this.client = client;
  }
  _name = "bot_chat_join_request";
  /** Approve the request */
  approve() {
    return this.client.hideJoinRequest({
      action: "approve",
      user: this.user.inputPeer,
      chatId: this.chat.inputPeer
    });
  }
  /** Decline the request */
  decline() {
    return this.client.hideJoinRequest({
      action: "decline",
      user: this.user.inputPeer,
      chatId: this.chat.inputPeer
    });
  }
}
exports.ChatJoinRequestUpdateContext = ChatJoinRequestUpdateContext;
