// @mtcute/dispatcher/context/message.js
import { Message, MtPeerNotFoundError } from "@mtcute/core";
class MessageContext extends Message {
  constructor(client, message) {
    const msg = Array.isArray(message) ? message[message.length - 1] : message;
    super(msg.raw, msg._peers, msg.isScheduled);
    this.client = client;
    this.messages = Array.isArray(message) ? message.map((it) => new MessageContext(client, it)) : [this];
    this.isMessageGroup = Array.isArray(message);
  }
  // this is primarily for proper types in filters, so don't bother much with actual value
  _name = "new_message";
  /**
   * List of messages in the message group.
   *
   * For other updates, this is a list with a single element (`this`).
   */
  messages;
  /** Whether this update is about a message group */
  isMessageGroup;
  /**
   * Get complete information about {@link sender}
   *
   * Learn more: [Incomplete peers](https://mtcute.dev/guide/topics/peers.html#incomplete-peers)
   */
  async getCompleteSender() {
    if (!this.sender.isMin) return this.sender;
    let res;
    if (this.sender.type === "user") {
      res = await this.client.getUser(this.sender);
    } else {
      res = await this.client.getChat(this.sender);
    }
    if (!res) throw new MtPeerNotFoundError("Failed to fetch sender");
    Object.defineProperty(this, "sender", { value: res });
    return res;
  }
  /**
   * Get complete information about {@link chat}
   *
   * Learn more: [Incomplete peers](https://mtcute.dev/guide/topics/peers.html#incomplete-peers)
   */
  async getCompleteChat() {
    if (!this.chat.isMin) return this.chat;
    let res;
    if (this.chat.type === "user") {
      res = await this.client.getUser(this.chat);
    } else {
      res = await this.client.getChat(this.chat);
    }
    if (!res) throw new MtPeerNotFoundError("Failed to fetch chat");
    Object.defineProperty(this, "chat", { value: res });
    return res;
  }
  /** Get a message that this message is a reply to */
  getReplyTo() {
    return this.client.getReplyTo(this);
  }
  /** If this is a channel post, get its automatic forward in the discussion group */
  getDiscussionMessage() {
    return this.client.getDiscussionMessage({ chatId: this.chat.inputPeer, message: this.id });
  }
  /** Get all custom emojis contained in this message (message group), if any */
  getCustomEmojis() {
    return this.client.getCustomEmojisFromMessages(this.messages);
  }
  /** Send a text message to the same chat (and topic, if applicable) as a given message */
  answerText(...params) {
    return this.client.answerText(this, ...params);
  }
  /** Send a media to the same chat (and topic, if applicable) as a given message */
  answerMedia(...params) {
    return this.client.answerMedia(this, ...params);
  }
  /** Send a media group to the same chat (and topic, if applicable) as a given message */
  answerMediaGroup(...params) {
    return this.client.answerMediaGroup(this, ...params);
  }
  /** Send a text message in reply to this message */
  replyText(...params) {
    return this.client.replyText(this, ...params);
  }
  /** Send a media in reply to this message */
  replyMedia(...params) {
    return this.client.replyMedia(this, ...params);
  }
  /** Send a media group in reply to this message */
  replyMediaGroup(...params) {
    return this.client.replyMediaGroup(this, ...params);
  }
  /** Send a text message in reply to this message */
  quoteWithText(params) {
    return this.client.quoteWithText(this, params);
  }
  /** Send a media in reply to this message */
  quoteWithMedia(params) {
    return this.client.quoteWithMedia(this, params);
  }
  /** Send a media group in reply to this message */
  quoteWithMediaGroup(params) {
    return this.client.quoteWithMediaGroup(this, params);
  }
  /** Send a text as a comment to this message */
  commentText(...params) {
    return this.client.commentText(this, ...params);
  }
  /** Send a media as a comment to this message */
  commentMedia(...params) {
    return this.client.commentMedia(this, ...params);
  }
  /** Send a media group as a comment to this message */
  commentMediaGroup(...params) {
    return this.client.commentMediaGroup(this, ...params);
  }
  /** Delete this message (message group) */
  delete(params) {
    return this.client.deleteMessagesById(
      this.chat.inputPeer,
      this.messages.map((it) => it.id),
      params
    );
  }
  /** Pin this message */
  pin(params) {
    return this.client.pinMessage({
      chatId: this.chat.inputPeer,
      message: this.id,
      ...params
    });
  }
  /** Unpin this message */
  unpin() {
    return this.client.unpinMessage({
      chatId: this.chat.inputPeer,
      message: this.id
    });
  }
  /** Edit this message */
  edit(params) {
    return this.client.editMessage({
      chatId: this.chat.inputPeer,
      message: this.id,
      ...params
    });
  }
  /** Forward this message (message group) */
  forwardTo(params) {
    return this.client.forwardMessagesById({
      fromChatId: this.chat.inputPeer,
      messages: this.messages.map((it) => it.id),
      ...params
    });
  }
  /** Send a copy of this message (message group) */
  copy(params) {
    if (this.isMessageGroup) {
      return this.client.sendCopyGroup({
        messages: this.messages,
        ...params
      });
    }
    return this.client.sendCopy({
      message: this,
      ...params
    });
  }
  /** React to this message */
  react(params) {
    return this.client.sendReaction({
      chatId: this.chat.inputPeer,
      message: this.id,
      ...params
    });
  }
}
export {
  MessageContext
};
