// @mtcute/dispatcher/context/scene-transition.cjs
if (typeof globalThis !== "undefined" && !globalThis._MTCUTE_CJS_DEPRECATION_WARNED) {
  globalThis._MTCUTE_CJS_DEPRECATION_WARNED = true;
  console.warn("[@mtcute/dispatcher] CommonJS bundles are deprecated. They will be removed completely in one of the upcoming releases. No support is provided for CommonJS users. Please consider switching to ESM, it's " + (/* @__PURE__ */ new Date()).getFullYear() + " already.");
  console.warn("[@mtcute/dispatcher] Learn more about switching to ESM: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c");
}
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const core = require("@mtcute/core");
const utils_js = require("@mtcute/core/utils.js");
const businessMessage = require("./business-message.cjs");
const callbackQuery = require("./callback-query.cjs");
const message = require("./message.cjs");
class SceneTransitionContext {
  constructor(previousScene, update) {
    this.previousScene = previousScene;
    this.update = update;
  }
  /** Get {@link update}, asserting it is a message-related update */
  get message() {
    if (this.update instanceof message.MessageContext) {
      return this.update;
    }
    throw new core.MtTypeAssertionError("SceneTransitionContext.message", "message", this.update._name);
  }
  /** Get {@link update}, asserting it is a business message-related update */
  get businessMessage() {
    if (this.update instanceof businessMessage.BusinessMessageContext) {
      return this.update;
    }
    throw new core.MtTypeAssertionError("SceneTransitionContext.businessMessage", "business message", this.update._name);
  }
  /** Get {@link update}, asserting it is a callback query update */
  get callbackQuery() {
    if (this.update instanceof callbackQuery.CallbackQueryContext) {
      return this.update;
    }
    throw new core.MtTypeAssertionError("SceneTransitionContext.callbackQuery", "callback query", this.update._name);
  }
  /** Get {@link update}, asserting it is an inline callback query update */
  get inlineCallbackQuery() {
    if (this.update instanceof callbackQuery.InlineCallbackQueryContext) {
      return this.update;
    }
    throw new core.MtTypeAssertionError(
      "SceneTransitionContext.inlineCallbackQuery",
      "inline callback query",
      this.update._name
    );
  }
}
utils_js.makeInspectable(SceneTransitionContext);
exports.SceneTransitionContext = SceneTransitionContext;
